transactions: allow disable of kafka_internal/tx

joe-redpanda opened the pull request:
Allow the transaction topic to be disabled.

Startup of the transaction partitions might crash the broker. Unlike other topics, tx previously could not be disabled. This can cause friction during livesite events where tx can continually crash a broker on startup.

Adds a simple test to enable and disable tx, validating that transactions can survive a disable/enable boundary and that transactions behave normally after reenable.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
### Improvements
* allow tx topic to be disabled if needed by operations

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

dotnwat commented on src/v/model/namespace.h:
```diff
@@ -98,6 +98,10 @@ inline const std::array non_user_topics{
   transform_log_internal_topic,
 };
 
+inline bool is_transaction_topic(topic_namespace_view tp_ns) {
```
this is unused

dotnwat commented on src/v/model/namespace.h:
```diff
@@ -107,9 +111,12 @@ inline bool is_user_topic(const ntp& ntp) {
 inline bool is_consumer_offsets_topic(const ntp& ntp) {
-    return ntp.ns == kafka_namespace
-           && ntp.tp.topic == kafka_consumer_offsets_topic;
+    return topic_namespace_view{ntp} == kafka_consumer_offsets_nt;
```
why did this change?

joe-redpanda replied:
I added operator == to topic_namespace_view so we don't have to copy paste
```
namespace == other_namespace && topic == other_topic
```

This is just cleanup to take advantage of operator==

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--78b11702-c978-4a0d-96d4-3bb3d198bf81-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71199">build#71199</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:---------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71199#0198cf23-109c-4730-985f-2a04033fa8e0 | FLAKY         | 19/21    | upstream reliability is '99.11111111111111'. current run reliability is '90.47619047619048'. drift is 8.63492 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

bharathv commented on src/v/model/namespace.h:
```diff
@@ -107,9 +111,12 @@ inline bool is_user_topic(const ntp& ntp) {
     return is_user_topic(topic_namespace_view{ntp});
 }
 
+inline bool can_be_disabled(topic_namespace_view tp_ns) {
```
nit: I think this doesn't belong here in namespace.h since "disabling" is cluster level concept. Perhaps move to somewhere this is used.

bharathv commented on tests/rptest/transactions/tx_disable_topic.py:
```diff
@@ -0,0 +1,230 @@
+        self.admin.set_partitions_disabled("kafka_internal", "tx")
+        self.logger.debug("Disabled kafka_internal/tx")
+
+        sleep(5)
```
use a conditional wait?  wait_until(foo...)

joe-redpanda replied:
fair, let me search for an api that confirms disabled

bharathv commented on tests/rptest/transactions/tx_disable_topic.py:
```diff
@@ -0,0 +1,230 @@
+        ''' checks if kafka_internal/tx can be disabled and reenabled
+            1. transaction remains valid between disable and reenable
+                a. start transaction
+                b. disable tx
```
what would init_transactions() return when disabled? retryable ec?

joe-redpanda replied:
let me add that, should be a cheap change

joe-redpanda replied:
init, produce, publish, commit all have timeouts and as a result just timeout on retryable

begin_transaction doesn't have a timeout. I'm not sure if this is significant or simply a client oddity.

bharathv replied:
A timeout doesn't seem right? Wondering if we should explicitly return COORDINATOR_NOT_AVAILABLE? Check it's usages [here](https://github.com/apache/kafka/blob/511818e027bbfc5eecc7b3d53dbb7a833dea5bd2/clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java) in reference Java client impl..

For regular produce and consume paths we detect if the partition is disabled and return replica_not_available ec.  I don't think Kafka has an equivalent disabling feature but  `COORDINATOR_NOT_AVAILABLE` seems more explicit IMO

joe-redpanda replied:
the broker is returning coordinator_not_available but confluent_kafka's client translates that to a timeout when your timer runs out

ck offers 'is_retriable' and 'txn_requires_abort' as the way to surface error details and we just check that these are 'yes retriable' and 'no abort needed'.

bharathv replied:
> the broker is returning coordinator_not_available but confluent_kafka's client translates that to a timeout when your timer runs out

sweet thanks for checking.

bharathv commented on tests/rptest/transactions/tx_disable_topic.py:
```diff
@@ -0,0 +1,230 @@
+        self.admin.set_partitions_disabled("kafka_internal", "tx", value=False)
+        self.logger.debug("Enabled kafka_internal/tx")
+
+        sleep(5)
```
same as above

bharathv commented on tests/rptest/transactions/tx_disable_topic.py:
```diff
@@ -0,0 +1,230 @@
+        blocked_producer.commit_transaction(5)
+        self.logger.debug("Blocked transaction commit completed successfully")
+
+        found_payloads = consume(3)
```
whats the issue with using get_transactions() again? Why do we have to consume and validate?

joe-redpanda replied:
get transactions doesn't show an active transaction after shutdown if the transaction spanned the disable -> enable gap

new transactions after the gap will though

let me add a snippet to demonstrate

joe-redpanda replied:
Not repro'ing today. This might have been an artifact of not sufficiently waiting for disable / reenable?

I'm not terribly worried because this feature is an esacape hatch. Its probably not the end of the world if the admin apis don't return accurate data intermittently while using an escape hatch

joe-redpanda committed:
model: add operator== to topic_namesapce_view

As title describes, adds equality operator to topic_namespace_view.
Changes model/metadata.h to make use of the new operator.

joe-redpanda committed:
cluster/topic_rules: add class for topic rules

Currently contains the rule for whether or not a topic can be disabled.

joe-redpanda committed:
cluster/topic_frontend: allow tx topic disable

Startup of the transaction controller partitions has previously crashed
brokers. Unlike other topics, tx previously could not be disabled. This
can cause friction during a livesite event.

The commit allows disabling kafka_internal/tx as an escape hatch for
operations.

joe-redpanda committed:
tests/disable_tx_topic: test tx disable behavior

Adds an integration test to check good behavior of the transaction apis
before, during, and after disable of the transaction controller topic.

bharathv commented on src/v/cluster/topic_rules.h:
```diff
@@ -0,0 +1,26 @@
+// contain rules for topic operations in an easy-to-find place
+class topic_rules {
+public:
+    static inline bool can_be_disabled(model::topic_namespace_view tp_ns) {
+        return model::is_user_topic(tp_ns) || (tp_ns == model::tx_manager_nt);
+    }
```
nit: a separate lib for this probably a bit too heavy, just make it a function in topics_frontend..cc?

joe-redpanda replied:
My thought was maybe there's someone else that is interested in whether a topic can be disabled or not (tests definitely, other prod code maybe)

I definitely want this to be more visible than a private method but i couldn't find a utils file that met the needs.

Thoughts?

bharathv replied:
ok cool, I can't think of an existing util to place this if not topics_frontend.

bharathv commented on tests/rptest/transactions/tx_disable_topic.py:
```diff
@@ -0,0 +1,365 @@
+    # change kafka_internal/tx state to enabled or disabled, wati until
+    # visible on all brokers
+    def change_tx_topic_state(self, should_disable=True):
+        log_prefix = 'Disabl'
```
:D
